### ISSUE TITLE
Play Order rework

### ISSUE THREAD

**Issue by lazka** [quodlibet/quodlibet#1411]
Original [issue 1411](https://code.google.com/p/quodlibet/issues/detail?id=1411) created by reiter.christoph on 2014-06-06T18:19:31.000Z:

Meta bug for play order changes...

Current design: http://f.666kb.com/i/cozwusc91t7xzr57s.png

* Two modes, shuffle/inorder
* Menu to select a modifier(check item) e.g. follow cursor / variant(radio item) e.g. weighted random

We can implement that on top of the current interface by grouping all orders in two groups... a good first step I guess. For the things below we probably need two kinds of plugins, orders and order wrappers.

Feedback welcome.

Menu examples:
```
--------------------------- <- modifiers
| [x] follow cursor (falls back to variant if cursor doesn't change)
| [ ] queue only (only affects clicks)
| [ ] autofill queue
| [ ] one song (falls back to variant on explicit track changes)
| [ ] track repeat (submenu to select count?) (falls back to variant on explicit track changes)
|-------------------------- <- variants
| o weigthed (random only)
| o spotify (random only)
| o reverse (inorder only)
---------------------------
```

The queue could have:

```
----------
| [x] stop after empty
| [ ] consume songs (remove after played)
| [ ] shuffle (or a menu button like in the main UI but only with variants, to keep things simple)
----------
```

Other bugs which should be considered:

* Issue #1407 - Spotify shuffle
* Issue #1111 - Repeat-block based play mode modifier
* Issue #903 - &quot;Next&quot; should not shuffle a new song if within a string of already played songs
* Issue #894 - Add &quot;party shuffle&quot; play order
* Issue #703 - Support un-shuffleable tracks for DJ mixes / classical pieces

Queue related:

* Issue #442 - The Play Queue should use the Play Order setting
* Issue #43 - rfe: stop playing once queue runs out

Implementation related:

* Issue #575 - Play order plugins should handle song list changes

**Comment by timpulver** [comment]
Very hard to follow this discussion and bring in new ideas. Could you post a summary of the problem and what might be planned?

**Comment by declension** [comment]
The closer I get to a refactoring path the more Kafkaesque the codebase seems. So some collected thoughts / plan:

### Highlights
* No complete rethink right now. The good ideas around pre-filling a queue e.g.  #1407, #904 would make this far too big to take on right now unfortunately.
* However, enable more powerful and flexible combinations of repeat and shuffle
* Simplify the code where possible. It's arguable the complexity here has meant that none of the abovementioned **11** issues have been closed yet (or maybe we're lazy, I don't know)

### Development plan

* [ ] Two widgets: shuffle (order) mode, repeat mode (both with on/off) - similar to current set up, but random has menu too (see old discussions above)
* [ ] Update prefs sections to allow these to be saved
* [ ] Consider migration issues in prefs as a result of above
* [ ] Some refactoring to cut through the layers / duplication in light of the above
* [ ] Make on/off less confusing, i.e. don't adjust the menu options dynamically.
* [ ] New `class Repeat(Order)` to model Repeat as an Order of its own separately. 
* [ ] A more flexible play system by using a single Order that is a composition with this new `Repeat` subclass wrapping, i.e. Repeat∘Shuffle. `None` (off state of widget) would be a pass-through for Repeat and for Shuffle could just be implemented as current `OrderInOrder` or equivalent. This makes more sense to me than any other scheme I considered, and would immediately solve #1111, #1836, plus a bunch of more niches uses (track repeat follow cursor, anyone?)
* [ ] New / updated `PluginHandler` to deal with plugin versions of `Repeat` orders. 
* [ ] Update current plugins to give more appropriate names given slightly new functionality 
* [ ] Convert existing plugins (e.g. TrackRepeat) to appropriate new types, and document.
* [ ] Rework / simplify `PlayOrder`, and possiubly `PlaylistModel`, arguably because details of play order and repeating has leaked out over our domain. So perhaps try avoid state in these, and by inverting control...
* [ ] ...so move single source of play controls nearer the origin, i.e. dependency of `Player`. For example, `BasePlayer.next()` currently delegates to `PlaylistModel.next()` which delegates to `Order.next_explicit()`, yet `Order` needs a back-reference to `PlaylistModel` in all of these requests.

**Comment by declension** [comment]
Lots of progress in [play-order branch](https://github.com/quodlibet/quodlibet/tree/play-order-updates).

**Comment by declension** [comment]
#2043 Merged

**Comment by NetherKing1357** [comment]
**This is a feature request.**

I've found this thread after a great deal of searching.
After spending some time figuring out the difference between the shuffle button at the bottom and the random option in the queue tab, I spent hours trying to get the queue to behave the way I want, with no success.

This is what I want from the queue:

1. Add from a list of songs to the queue, while shuffling them (thus generating a shuffled queue)
2. Observe upcoming tracks in the queue tab
3. Have the ability to change the order of upcoming tracks
4. Make this the default mode of shuffle

Is there any method of doing this right now? If not, is it in the works?

**Difficulty of implementing:**
Unknown
